utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,248 @@
+///         let ret = mem::transmute_copy(&from);
+///         mem::forget(from);
+///         ret
+///     }
+///
+///     fn with_mut<F>(&'a mut self, f: F)
+///     where
+///         F: 'static + FnOnce(&'a mut Self::Output),
+///     {
+///         unsafe { f(mem::transmute::<&mut Self, &mut Self::Output>(self)) }
+///     }
+/// }
+/// ```
+pub unsafe trait Yokeable<'a>: 'static {
+    /// This type MUST be `Self` with the `'static` replaced with `'a`, i.e. `Self<'a>`
+    type Output: 'a + Sized;
+
+    /// This method must cast `self` between `&'a Self<'static>` and `&'a Self<'a>`.
+    ///
+    /// # Implementation safety
+    ///
+    /// If the invariants of [`Yokeable`] are being satisfied, the body of this method
+    /// should simply be `{ self }`, though it's acceptable to include additional assertions
+    /// if desired.
+    fn transform(&'a self) -> &'a Self::Output;
+
+    /// This method can be used to cast away `Self<'a>`'s lifetime.
+    ///
+    /// # Safety
+    ///
+    /// The returned value must be destroyed before the data `from` was borrowing from is.
+    ///
+    /// # Implementation safety
+    ///
+    /// A safe implementation of this method must be equivalent to a transmute between
+    /// `Self<'a>` and `Self<'static>`
+    unsafe fn make(from: Self::Output) -> Self;
+
+    /// This method must cast `self` between `&'a mut Self<'static>` and `&'a mut Self<'a>`,
+    /// and pass it to `f`.
+    ///
+    /// # Implementation safety
+    ///
+    /// A safe implementation of this method must be equivalent to a pointer cast/transmute between
+    /// `&mut Self<'a>` and `&mut Self<'static>` being passed to `f`
+    ///
+    /// # Why is this safe?
+    ///
+    /// Typically covariant lifetimes become invariant when hidden behind an `&mut`,
+    /// which is why the implementation of this method cannot just be `f(self)`.
+    /// The reason behind this is that while _reading_ a covariant lifetime that has been cast to a shorter
+    /// one is always safe (this is roughly the definition of a covariant lifetime), writing
+    /// may not necessarily be safe since you could write a smaller reference to it. For example,
+    /// the following code is unsound because it manages to stuff a `'a` lifetime into a `Cow<'static>`
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// struct Foo {
+    ///     str: String,
+    ///     cow: Cow<'static, str>,
+    /// }
+    ///
+    /// fn unsound<'a>(foo: &'a mut Foo) {
+    ///     let a: &str = &foo.str;
+    ///     foo.cow.with_mut(|cow| *cow = Cow::Borrowed(a));
+    /// }
+    /// ```
+    ///
+    /// However, this code will not compile because [`Yokeable::with_mut()`] requires `F: 'static`.
+    /// This enforces that while `F` may mutate `Self<'a>`, it can only mutate it in a way that does
+    /// not insert additional references. For example, `F` may call `to_owned()` on a `Cow` and mutate it,
```

Comment:
Are you sure? What if it accesses a thread-local?

I don't believe this is an issue but it's worth explicitly checking this case. (Also, I wonder if there is evil to be done with global mutices...)

Reply:
You can't _borrow_ from a thread local or a global mutex, you can only do scoped operations on them.

Reply:
Oh, what about
```
let empty: &'static mut [u8] = &mut [];
```

Similarly, `Box::leak`? I don't *think* these should be an issue either, because those are truly `'static` and that lifetime is covariant. Just trying to nail down every source of borrowed data I can think of.

Reply:
Yeah, those are truly `'static`. And static mut is unsafe to work with anyway.